integration/src/main/java/scripts/testscripts/DatasetPermissionOperations.java

```diff
@@ -0,0 +1,238 @@
+package scripts.testscripts;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.fail;
+
+import bio.terra.catalog.api.DatasetsApi;
+import bio.terra.catalog.client.ApiException;
+import bio.terra.catalog.model.CreateDatasetRequest;
+import bio.terra.catalog.model.StorageSystem;
+import bio.terra.datarepo.api.JobsApi;
+import bio.terra.datarepo.api.SnapshotsApi;
+import bio.terra.datarepo.model.JobModel;
+import bio.terra.datarepo.model.PolicyMemberRequest;
+import bio.terra.datarepo.model.SnapshotRequestContentsModel;
+import bio.terra.datarepo.model.SnapshotRequestModel;
+import bio.terra.testrunner.runner.TestScript;
+import bio.terra.testrunner.runner.config.TestUserSpecification;
+import com.google.api.client.http.HttpStatusCodes;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scripts.client.CatalogClient;
+import scripts.client.DatarepoClient;
+
+public class DatasetPermissionOperations extends TestScript {
+
+  private static final Logger log = LoggerFactory.getLogger(DatasetPermissionOperations.class);
+  private static final String TEST_DATASET_NAME = "CatalogTestDataset";
+
+  private TestUserSpecification adminUser;
+  private TestUserSpecification regularUser;
+
+  // TDR APIs
+  private SnapshotsApi adminSnapshotsApi;
+  private List<UUID> snapshotIds = new ArrayList<>();
+  private UUID defaultProfileId;
+
+  // Catalog APis
+  private List<UUID> datasetIds = new ArrayList<>();
+  private CatalogClient adminClient;
+  private DatasetsApi adminDatasetsApi;
+
+  @Override
+  public void setup(List<TestUserSpecification> testUsers) throws Exception {
+    for (TestUserSpecification testUser : testUsers) {
+      if ("datacatalogadmin@test.firecloud.org".equals(testUser.userEmail)) {
+        adminUser = testUser;
+      }
+      if ("datacataloguser@test.firecloud.org".equals(testUser.userEmail)) {
+        regularUser = testUser;
+      }
+    }
+    var datarepoClient = new DatarepoClient(server, adminUser);
+    adminSnapshotsApi = new SnapshotsApi(datarepoClient);
+    defaultProfileId =
+        new bio.terra.datarepo.api.DatasetsApi(datarepoClient)
+            .enumerateDatasets(null, null, null, null, TEST_DATASET_NAME, null)
+            .getItems()
+            .get(0)
+            .getDefaultProfileId();
+    adminClient = new CatalogClient(server, adminUser);
+    adminDatasetsApi = new DatasetsApi(adminClient);
+  }
+
+  @Override
+  public void userJourney(TestUserSpecification testUser) throws Exception {
+    DatasetsApi userDatasetsApi = new DatasetsApi(new CatalogClient(server, regularUser));
+
+    testReaderPermissions(userDatasetsApi);
```

Comment:
Can the same snapshot can be used for all of these tests, since they aren't running in parallel? Each test would have to ensure that the permissions are in the expected state before running, though. Since snapshots are so expensive to make, it would be worth recycling one if possible.